src/stringToPath.test.ts

```diff
@@ -1,11 +1,68 @@
 import { stringToPath } from "./stringToPath";
 
-test("should convert a string to a deeply nested path", () => {
-  const res = stringToPath("a.b[0].c");
-  expect<["a", "b", "0", "c"]>(res).toEqual(["a", "b", "0", "c"]);
-});
+describe("stringToPath", () => {
+  describe("implementation", () => {
+    test("should convert a string to a deeply nested path", () => {
+      const res = stringToPath("foo.bar[3].baz");
+      expect(res).toEqual(["foo", "bar", "3", "baz"]);
+    });
+
+    test("should handle nested dot paths", () => {
+      const res = stringToPath("foo[bar.baz].qui.che");
+      expect(res).toEqual(["foo", "bar.baz", "qui", "che"]);
+    });
+
+    test("should handle short path with only bracket access", () => {
+      const res = stringToPath("foo[bar]");
+      expect(res).toEqual(["foo", "bar"]);
+    });
+
+    test("should handle bracket access at the end", () => {
+      const res = stringToPath("foo.bar[3]");
+      expect(res).toEqual(["foo", "bar", "3"]);
+    });
+
+    test("should handle relatively long paths without performance issues", () => {
+      const res = stringToPath(
+        "lorem.ipsum[dolor.sit].amet.con.sec.tetur[adi.pisc.ing].elit.42",
+      );
+      expect(res).toEqual([
+        "lorem",
+        "ipsum",
+        "dolor.sit",
+        "amet",
+        "con",
+        "sec",
+        "tetur",
+        "adi.pisc.ing",
+        "elit",
+        "42",
+      ]);
+    });
+  });
+
+  describe("types", () => {
+    test("simple const string are inferred", () => {
+      const input = "foo[bar.baz].qui";
+      const data = stringToPath(input);
+      assertType<["foo", "bar.baz", "qui"]>(data);
+    });
+
+    test("string constructed from const strings are inferred", () => {
+      const bar = "bar";
+      const baz = "baz";
+      const qui = "qui";
+      const input = `foo[${bar}.${baz}].${qui}.che`;
```

Comment:
This isn't testing what you think it's testing, it's equivalent to just a const string, typescript doesn't do anything different with this type:

https://www.typescriptlang.org/play?#code/MYewdgzgLgBARgQwE4wLwwESKRg3AKFElkQC81My9DxoYBHAVwEsKMnnqi7mwAHRrHQADAGYgQAbQAkAb2wBfAHRyyCgLorZHZcAAWAU2EEA9CZgWYAPQD8+GsRi8BUAEwUxEydiVlNHJX0jAnwoAE8+AxgADQpwyJBRJ35BGAMADygDMAATCBh4g0Tkl1d8M0tbfAsbAoiipOdBdwys3PzC4qaoastaqCRGA16LAC4YUQQAGwhhsYnp2YIgA


```suggestion
      const input = `foo[bar.baz].qui.che`;
```

Reply:
It's testing exactly what I wanted to test, see https://github.com/remeda/remeda/pull/630#discussion_r1560139684 - I know it seems kind of like an obvious test-case, but I think it doesn't hurt to have it explicitly in the suite - shows we've thought of the case?

Reply:
This test and the test above it test the same thing, we don't need to test that typescript can't infer literal strings, that's on the typescript team to do. There are no cases where this test would break and the previous test wouldn't, and vice-versa. In that regard the syntax here is confusing as it looks like something it isn't, and might throw off a maintainer in the future.